microceph/ceph/osd.go

```diff
@@ -438,28 +439,413 @@ func AddOSD(s *state.State, path string, wipe bool, encrypt bool) error {
+
+	for i := 0; i < retries; i++ {
+		err = doPurge(osd)
+		if err == nil {
+			// Success: break the retry loop
+			break
+		}
+		// we're getting a RunError from processExec.RunCommand, and it
+		// wraps the original exit error if there's one
+		exitError, ok := err.(shared.RunError).Unwrap().(*exec.ExitError)
+		if !ok {
+			// not an exit error, abort and bubble up the error
+			logger.Warnf("Purge failed with non-exit error: %v", err)
+			break
+		}
+		if syscall.Errno(exitError.ExitCode()) != syscall.EBUSY {
+			// not a busy error, abort and bubble up the error
+			logger.Warnf("Purge failed with unexpected exit error: %v", exitError)
+			break
+		}
+		// purge failed with EBUSY - retry after a delay, and make delay exponential
+		logger.Infof("Purge failed %v, retrying in %v", err, backoff)
+		backoff = time.Duration(math.Pow(2, float64(i))) * time.Millisecond * 100
+		time.Sleep(backoff)
+	}
+
+	if err != nil {
+		logger.Errorf("Failed to purge osd.%d: %v", osd, err)
+		return fmt.Errorf("Failed to purge osd.%d: %w", osd, err)
+	}
+	logger.Infof("osd.%d purged", osd)
+	return nil
+}
+
+func wipeDevice(s common.StateInterface, osd int64) error {
+	// get the device path
+	path, err := database.OSDQuery.Path(s.ClusterState(), osd)
+	if err != nil {
+		return err
+	}
+	// wipe the device, retry with exponential backoff
+	retries := 8
+	var backoff time.Duration
+	for i := 0; i < retries; i++ {
+		err = timeoutWipe(path)
+		if err == nil {
+			// Success: break the retry loop
+			break
+		}
+		// wipe failed - retry after a delay, and make delay exponential
+		logger.Infof("Wipe failed %v, retrying in %v", err, backoff)
+		backoff = time.Duration(math.Pow(2, float64(i))) * time.Millisecond * 100
+		time.Sleep(backoff)
+	}
+	if err != nil {
+		// log a warning, but don't treat wipe failure as a fatal error
+		// e.g. if the device is broken, we still want to remove it from the cluster
+		logger.Warnf("Fault during device wipe: %v", err)
+	}
+	return nil
+}
+
+// timeoutWipe wipes the given device with a timeout, in order not to hang on broken disks
+func timeoutWipe(path string) error {
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
+
+	_, err := processExec.RunCommandContext(
+		ctx,
+		"dd", "if=/dev/zero",
+		fmt.Sprintf("of=%s", path),
+		"bs=4M", "count=10", "status=none",
+	)
+	return err
+}
+
+func doRemoveOSD(ctx context.Context, s common.StateInterface, osd int64, bypassSafety bool) error {
+	var err error
+
+	// general sanity
+	err = sanityCheck(s, osd)
+	if err != nil {
+		return err
+	}
+
+	if !bypassSafety {
+		// check: at least 3 OSDs
+		err = checkMinOSDs(s, osd)
 		if err != nil {
-			return fmt.Errorf("Failed to fetch disks: %w", err)
+			return err
+		}
+	}
+
+	err = scaleDownFailureDomain(s, osd)
+	if err != nil {
+		return err
+	}
+
+	// check if the osd is still in the cluster -- if we're being re-run, it might not be
+	isPresent, err := haveOSDInCeph(osd)
```

Comment:
This should probably be moved a bit more up; there's no reason to scale down _before_ checking that the OSD exists in the DB, IMO.

Reply:
The scale down is designed to be independent of any individual osd, it will determine the need to downgrade solely based on cluster state, so it should not matter if it comes before or after the presence check. 
I have opted to run the check for downscaling before the presence check for robustness, if for some reason (e.g. we're being re-run and there was some fault previously) the OSD already was taken out, we get another chance to downscale